python/mspasspy/db/normalize.py

```diff
@@ -0,0 +1,2052 @@
+                        imin = i
+                        dtmin = dt
+                return matchlist[imin][1]
+
+    def normalize(self, d):
+        if d.dead():
+            return d
+        if _input_is_atomic(d):
+            doc = self.get_document(d)
+            if doc == None:
+                message = (
+                    "No matching document was found in"
+                    + self.arrival_collection_name
+                    + " collection for this datum"
+                )
+                self.log_error(
+                    d,
+                    "css30_arrival_interval_matcher",
+                    message,
+                    self.kill_on_failure,
+                    ErrorSeverity.Invalid,
+                )
+            else:
+                for key in self.attributes_to_load:
+                    if key in doc:
+                        if self.prepend_collection_name:
+                            mdkey = self.collection + "_" + key
+                        else:
+                            mdkey = key
+                        d[mdkey] = doc[key]
+                    else:
+                        # We accumulate error messages to aid user debugging
+                        # but it could create bloated elog collections
+                        message = (
+                            "No data for key="
+                            + self.mdkey
+                            + " in document returned from collection="
+                            + self.collection
+                        )
+                        self.log_error(
+                            d,
+                            "css30_arrival_interval_matcher",
+                            message,
+                            self.kill_on_failure,
+                            ErrorSeverity.Invalid,
+                        )
+                # similar for optional but don't log errors for missing
+                # attributes unless verbose is set true
+                for key in self.load_if_defined:
+                    if key in doc:
+                        if self.prepend_collection_name:
+                            mdkey = self.collection + "_" + key
+                        else:
+                            mdkey = key
+                        d[mdkey] = doc[key]
+                    elif self.verbose:
+                        self.log_error(
+                            "css30_arrival_interval_matcher",
+                            "No data found with optional load key=" + key,
+                            ErrorSeverity.Informational,
+                        )
+
+            # Notice logic that if no match is found we log the error
+            # (and usually kill it) and land here. Allows application in a map
+            # operation
+            return d
+        else:
+            raise TypeError(
+                "css30_arrival_interval_matcher.normalize:  received invalid data type"
+            )
+
+
+def bulk_normalize(
```

Comment:
You extended this to be much more generic.   Has it been tested on anything but the default?   Docstring needs some clarification on usage.   My prototype was frozen to only handle what here is the default.

Reply:
I looked at this again this morning and realized I misunderstood what this class was doing.   It is not a generic version of normalize_mseed but a method to do apply mulitple normalizers in a single pass.  I retract my comments on this class and say instead the docstring needs to be expanded to better clarify its use.

Reply:
Hey Gary, I just read through your comments, thanks for pointing them out. I will add the missing docstrings. 

Also, in this week's meeting, Ian and I went through the code and decided to make some revisions. Since some different derived classes of NMF share lots of duplicated codes (for example, mseed_channel_matcher and mseed_site_matcher), It might be better to extract the duplicated parts, and put them in the NMF or another intermediatory class, to avoid the duplication and complexity. I am currently working on it, and I will push the update very soon.

Reply:
Makes a lot of sense.  That is a classic use of a base class.